pkg/ebpf/tracer.go

```diff
@@ -766,38 +772,48 @@ func (t *Tracer) DebugNetworkMaps() (*network.Connections, error) {
 // closed ports will be returned.
 // the map will be one of port_bindings  or udp_port_bindings, and the mapping will be one of tracer#portMapping
 // tracer#udpPortMapping respectively.
-func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) ([]uint16, error) {
-	var key, emptyKey uint16
+func (t *Tracer) populatePortMapping(mp *ebpf.Map, mapping *network.PortMapping) (closed []portBindingTuple, closedNetNs []uint64, err error) {
+	var key, emptyKey portBindingTuple
 	var state uint8
 
-	closedPortBindings := make([]uint16, 0)
-
+	inos := make(map[int]uint64)
 	entries := mp.IterateFrom(unsafe.Pointer(&emptyKey))
 	for entries.Next(unsafe.Pointer(&key), unsafe.Pointer(&state)) {
-		port := key
+		pid := int(key.pid)
+		port := uint16(key.port)
+		var ino uint64
+		var ok bool
+		if ino, ok = inos[pid]; !ok {
+			var err error
+			ino, err = util.GetNetNsInoFromPid(t.config.ProcRoot, pid)
```

Comment:
I'm worried about adding calls to procfs in the hot path. Can we get at the inode number in eBPF ?

Reply:
This is also going to be racy since there's no guarantee that the process is still alive/has a network namespace when   bound.

There is the get_task_info which requires kernel 4.8 that seems to have this information.  

Could we map tid/pid -> namespace in another kprobe ?

Reply:
I couldn't find any reference to a `get_task_info` function. Can you point me to where you saw that?

Reply:
https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#7-bpf_get_current_task

Reply:
FYI that helper is not available until 4.8

Reply:
Having a generic kprobe to track pid to net ns mappings would be challenging as network namespaces can be created multiple ways. There is only one instance where we couldn't get the the network namespace since we didn't have a `sock` object. Maybe we can use this in that instance.

Reply:
As discussed offline get_task_info is only available in 4.8+ so we can't rely on it. 

One other way I can think of to get around this is to have the kprobe for ip_mk_skb rewrite the UDP entry in the listening map since we have access to a `sock*` object in that kprobe.

Reply:
There is another problem with this approach. If we set `net_ns` to `0` for udp port mappings, when we resolve the pid to net ns, it may succeed, and we will add it to the port mappings with the net ns. However, when the port is closed, the pid is very likely gone also, and so we end up not removing the port mapping since we don't have the net ns. 

I think we should just remove this code for now. UDP port bindings will be broken in this case, but we can fix that by using `inet_bind` as I suggested in a different comment.